### ISSUE TITLE
Allow exporting root functions as a means of outputting parameterised configs

### ISSUE THREAD

**Issue by PatrickShaw** [tweag/nickel#1202]
**Is your feature request related to a problem? Please describe.**

Been experimenting with replacing YAML with Nickel and immediately ran into a situation where there's no easy means of reusing the same Nickel file.

I commonly have a case where there's 1 or 2 fields that need to differ between OS. For instance, my [Alacritty](https://github.com/alacritty/alacritty) YAML for Linux looks like:

```yaml
# ...some shared config
 window:
  opacity: 0.9
```

while my MacOS and other OS configs look like:

```yaml
# ...some shared config
window:
  opacity: 1
```

To match the aesthetic of the respective desktop environments that I have in those operating systems.

At the moment, to my knowledge, there's no way of parameterising Nickel configs

**Describe the solution you'd like**
I think there would be a lot of power in the ability to export `fun` as the root of a Nickel file.

To solve the original problem statement, you could write the following .ncl file:

```ncl
# alacritty.ncl
fun system => {
  # ...some shared config
  window = {
    opacity = if system == "linux" then 0.9 else 1
  }
}
```

and run:

```bash
nickel -f alacritty.yml --format yaml --input linux > ./linux/alacritty.yaml
nickel -f alacritty.yml --format yaml --input linux > ./macos/alacritty.yaml
```

Each positional argument could be declared with another `--input` flag.

**Describe alternatives you've considered**

My workaround is currently to use imports like so:

```ncl
# shared.ncl
fun system => {
  # ...some shared config
  window = {
    opacity = if system == "linux" then 0.9 else 1
  }
}
```

with

```ncl
# linux.ncl
import "./shared.ncl" "linux"
```

```ncl
# macos.ncl
import "./shared.ncl" "macos"
```

However, creating those files is a slight inconvenience and becomes more inconvenient in cases where there is more than one parameter with more than 2 variations to choose from.

**Additional context**
See https://github.com/PatrickShaw/my-monorepo/tree/7027b953de39e32a919dbb86ca8a4b3b7c1728d3/personal/dotfiles/alacritty for the real use case (i.e. Unsimplified example from above)

**Comment by matthew-healy** [comment]
Does something like this solve your problem?

```
# alacritty.ncl
{
  base = # ...shared config
  macos = base & { window.opacity = 1 },
  linux = base & { window.opacity = 0.9 },
}
```

And then...
```
echo '(import "alacritty.ncl").macos' | nickel export --format=yaml
```
or...
```
echo '(import "alacritty.ncl").linux' | nickel export --format=yaml
```

**Comment by yannham** [comment]
See also #413. However, whatever we chose to implement (or not) with respect to passing arguments, @matthew-healy 's suggestions stand, in that those CLI parameters would only be helping reducing verbosity, but they wouldn't provide anything that can't be done today already.

As an alternative, you can also push the customization into your command line invocation as well:

```nickel
# alacritty.ncl
{
  # your config... (Matthew's "base")
  window = {
    opacity | default = 1,
  }
}
```

```console
$ nickel export --format=yaml <<< 'import "alacritty.ncl" & {window.opacity = 0.5}'
..YAML OUTPUT..
$ nickel export --format=yaml <<< 'import "alacritty.ncl" & {window.opacity = 0.9}'
..YAML OUTPUT..
```

Yet another solution is to put them into separate files:

```nickel
# macos.ncl
import "alacritty.ncl" & {
  window.opacity  = 1.0
}

# linux.ncl
import "alacritty.ncl" & {
  window.opacity = 0.9
}
```

And then simply call `nickel export -f macos.ncl --format yaml`. Doing so your specific configuration file really looks like configuration: you're just overriding settings with XXX-specific value.

## A note on record and merging

You intuitively went for a function to parametrize your configuration, which totally makes sense. However, Nickel features recursive records and merging, which somehow lets you achieve the same but in a nicer way. You can see this technique as defining partial configurations (like `base`), with values potentially depending on each others. Then, the merge operation `&` lets you fill the missing values, or override some (using the right priorities). You're combining piece of configurations until reaching a definite result. Cf https://nickel-lang.org/user-manual/merging.

**Comment by PatrickShaw** [comment]
Thank you both @matthew-healy @yannham for your suggestions.

If I could potentially provide some additional perspective to supporting functions:

here-string/echo-piping makes niceties like syntax highlighting, autosuggestion and formatting for the inline Nickel code non-trivial as it requires syntax highlighters to understand both the shell syntax (bash/sh/etc in the case of the suggestions) and know that the string literal refers to nickel code
- If I understand correctly, , the [contract](https://github.com/tweag/nickel/blob/dcbbfedef46fe67d0507ad7a0d0326f985321e4a/doc/manual/contracts.md#by-hand)

They do both serve the same purpose as my proposal in that they avoid save the need to write files.

**Comment by PatrickShaw** [comment]
Thank you both @matthew-healy @yannham for your suggestions. I had not considered using stdin. I agree nothing in my proposal can't already be done. I also prefer these to my current solution :)

If I could add some additional perspective to supporting functions:

**IDE/Editor niceties**
I can see some benefit to avoiding here-strings/echo-piping in that they're harder to provide syntax highlighting, formatting and autosuggestion support for since they require knowledge that the string is Nickel code.

I admit the inlined code would, generally, be quite minimal.

**Dynamism and potential security attack surface**
Although I have not run into this situation myself, I can imagine that some parameters cannot be hard coded into the inline Nickel code. Secret keys would be an example of this.

Using @yannham 's solution as a simple example, we might change:

```bash
nickel export --format=yaml <<< 'import "alacritty.ncl" & {window.opacity = 0.5}'
```

to

```bash
#!/bin/bash
# example.sh
nickel export --format=yaml <<< 'import "alacritty.ncl" & {window.opacity = $OPACITY }'
```

however this introduces some concern over code injection, accidental or otherwise:
```bash
OPACITY='0.9 } & { DANGEROUS_PARAMETER="scary-value" ' ./example.sh
```

The [shell](https://github.com/alacritty/alacritty/blob/2df8f860b960d7c96efaf4f059fe2fbbdce82bcc/alacritty.yml#L454) field could be abused in this case, for instance, to execute arbitrary programs upon executing Alacritty.


Any thoughts?

**Comment by vkleen** [comment]
That's a really interesting consideration. On the one hand, we could punt on it and just say that injections like that are the command line users responsibility. On the other hand, it would be very useful to have a "blessed" way of overriding configuration fields from the command line, anyway.

For example, we could have a command line parameter `--override` that takes a field path and a Nickel term. Semantically, say for an invocation `nickel --override field.path.x "{ some = nickel, nickel = 5 }" file.ncl`, the interpreter would behave as if evaluating
```
import "file.ncl" & { field.path.x = { some = nickel, nickel = 5 } }
```

This would eliminate the injection concern, keep with recursive records as the method of overriding and be much nicer to use than piping things into stdin.

**Comment by matthew-healy** [comment]
I was thinking about whether we could avoid having to write any Nickel code outside an `ncl` file, by allowing you to specify a top-level field you'd like to export from a given record.

E.g. something like:

```bash
$ nickel export --format yaml --file ./alacritty.ncl --field linux
# equivalent to: nickel export --format yaml <<< '(import alacritty.ncl).linux'
```

This seems to me to be the cleanest way of solving this issue, without leaking any Nickel code to the command itself.

**Comment by yannham** [comment]
> Code injection attack surface would be smaller depending on the type contract of the function in my proposal.
> Any thoughts?

Beside a new feature, you can simply write the value `$OPACITY` in a separate Nickel file, which would prevent it from "escaping its scope" in any way. Doing so you would even have less code to write on the command line, as you know everything statically, but the value of `$OPACITY`:

```nickel
# file combine.ncl
import "alacritty.ncl" & {
  window.opacity = import "opacity".ncl,
}
```

```console
$ echo "$OPACITY" > opacity.ncl
$ nickel export -f combine.ncl --format yaml
```

**Comment by yannham** [comment]
> E.g. something like:
> 
> ```shell
> $ nickel export --format yaml --file ./alacritty.ncl --field linux
> # equivalent to: nickel export --format yaml <<< '(import alacritty.ncl).linux'
> ```
> 
> This seems to me to be the cleanest way of solving this issue, without leaking any Nickel code to the command itself.

I feel like you are answering a different (but valid!) question, which is when we want to export a specific field of a configuration. There is indeed a discussion asking for this (e.g #825).

However this issue is about setting the parameters of a configuration from the command-line (either as a function or a recursive record), but export everything, a priori.

**Comment by matthew-healy** [comment]
It's a different question in some respects, but at a certain point we have to ask what the utility of these features are. As previously discussed in the thread, specialising configuration by passing arguments to functions isn't necessarily "idiomatic Nickel". I'm not convinced that this is really a core use-case we should be supporting, and I think there are better options for solving this problem (e.g. by exporting subfields).

**Comment by yannham** [comment]
I still don't see the relation between exporting subfields and passing arguments. Assume you have such a `--field` option, how does it help solving the original example?

**Comment by matthew-healy** [comment]
>Assume you have such a --field option, how does it help solving the original example?

It doesn't, but as you said above, the initial example is probably better expressed with record merging like [in my comment above](https://github.com/tweag/nickel/issues/1202#issuecomment-1488157917). So my suggestion is to encourage that approach to this sort of problem, rather than exporting & calling top-level functions at all. 

IMO, that solves the overall use case of "how do I customise a piece of base configuration" without requiring a mechanism for passing args to Nickel functions from the command line, which is a feature I'm not entirely convinced we need.

**Comment by yannham** [comment]
Well, even with record merging, you do need to pass the new value for window opacity _in some way_, don't you? I think `--field` relates to the *output*. When our output is a record and we only want one part of it.

But this issue seems to be concerned with *input*. Even if you don't need the `.linux` at the end thanks to `--field`, whatever is your approach for customization - function or recursive record - you still need to merge the base config with a new value, or to apply it to the new value (here `$OPACITY`).

I guess you can then propose that we have an `--override` as @vkleen suggested, which indeed favor the record merging (which wouldn't need, or at least less often, to actually write code). But that would be quite similar to an `--arg` parameter. So we still need a special treatment for inputs in both cases if we don't want to write Nickel code on the command line, it's just given a different name and have different characteristics. Or am I missing something?

**Comment by yannham** [comment]
@PatrickShaw, can I consider this issue as a duplicate of #413, or do you think your request is different in some way?

**Comment by PatrickShaw** [comment]
Hi all, sorry I've been holidays for the past few weeks. 

@yannham yes let's mark this as a duplicate as I think they speak to the same use cases. I'll close the issue.

 I found this thread very useful. Thanks @yannham @vkleen @matthew-healy for the discussion.

Just going to add my 2c as some extra data for some of the ideas that were discussed.

---

> Beside a new feature, you can simply write the value $OPACITY

This didn't cross my mind and definitely solves the code injection concerns. Works well with contracts too if I pass the file's output straight into the input of a function.

It's also pretty easy to simply delete the file if we instead assume we need to pass in a secret so this is solved as well, although a little inconvenient.

I'm agreement now that this proposal is really just for to allow for convenience and does not really enable any new use case.

---

> we could have a command line parameter --override

> nickel export --format yaml --file ./alacritty.ncl --field linux

Regarding the `--override` and `--field` ideas. These would solve the opacity case, though I'm honestly quite happy with the already existing solutions to solving this case. I've personally settled for:

```nickel
  nickel export --format yaml <<< 'import "shared.ncl" "linux"' > ./linux/alacritty.yaml
  nickel export --format yaml <<< 'import "shared.ncl" "macos"' > ./macos/alacritty.yaml
```

which is enough for my purposes at the moment :) 

`--field` unfortunately would unfortunately still require storing the secret key in the second hypothetical case I proposed.

I'm not a huge fan `--override` in that I would, personally, like to treat the Nickel file itself as a sort of function that abstracts away how the final output was calculated. Having an `--override` at the output level would require that consumer of the Nickel file knows how the output is structured and how it's computed in cases where the output is complicated to compute. What I'm mostly interested in are ways of providing inputs into the Nickel file so I don't have to concern myself with the output.
